File: day5/part1.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_RANGES 10000
#define MAX_IDS 100000
#define LINE_LEN 128

typedef struct {
  int start;
  int end;
} Range;

bool contains(const Range *r, int value) {
  return value >= r->start && value <= r->end;
}

int main(void) {
  FILE *file = fopen("input.txt", "r");
  if (!file) {
    perror("Failed to open input.txt");
    return 1;
  }

  Range ranges[MAX_RANGES];
  int ids[MAX_IDS];

  int range_count = 0;
  int id_count = 0;
  bool parsing_ranges = true;

  char line[LINE_LEN];

  while (fgets(line, sizeof(line), file)) {
    line[strcspn(line, "\n")] = '\0';

    if (strlen(line) == 0) {
      parsing_ranges = false;
      continue;
    }

    if (parsing_ranges) {
      sscanf(line, "%d-%d", &ranges[range_count].start,
             &ranges[range_count].end);
      range_count++;
    } else {
      ids[id_count++] = atoi(line);
    }
  }

  fclose(file);

  int fresh_count = 0;

  for (int i = 0; i < id_count; i++) {
    for (int j = 0; j < range_count; j++) {
      if (contains(&ranges[j], ids[i])) {
        fresh_count++;
        break;
      }
    }
  }

  printf("%d\n", fresh_count);
  return 0;
}

File: day5/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Range struct {
	Start int
	End   int
}

func (r Range) Contains(value int) bool {
	return value >= r.Start && value <= r.End
}

func parseInput(path string) ([]Range, []int, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	var ranges []Range
	var ids []int
	scanner := bufio.NewScanner(file)

	parsingRanges := true

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			parsingRanges = false
			continue
		}

		if parsingRanges {
			parts := strings.Split(line, "-")
			start, _ := strconv.Atoi(parts[0])
			end, _ := strconv.Atoi(parts[1])
			ranges = append(ranges, Range{Start: start, End: end})
		} else {
			value, _ := strconv.Atoi(line)
			ids = append(ids, value)
		}
	}

	return ranges, ids, scanner.Err()
}

func countFreshIDs(ranges []Range, ids []int) int {
	count := 0
	for _, value := range ids {
		for _, r := range ranges {
			if r.Contains(value) {
				count++
				break
			}
		}
	}
	return count
}

func main() {
	ranges, ids, err := parseInput("input.txt")
	if err != nil {
		panic(err)
	}

	result := countFreshIDs(ranges, ids)
	fmt.Println(result)
}
